### PR TITLE
[Slider] Allow for different thumb radiuses for different states

### LINKED ISSUE #4883: [Slider] Support different Thumb radius for UIControlState
## Feature Request/Suggestion

The MDCSlider thumb should support changing size (radius) for each UIControlState value.

For example:
* Disabled: 2 points
* Normal: 6 points
* Selected: 10 points

---
#### Internal data
- Associated internal bug: [b/117176821](http://b/117176821)

### PR DESCRIPTION
**Pull request by codeman7**
This change allows clients to set different thumb radiuses for all different `UIControlState`s. This will match the existing API for trackTickColor, trackFillColor, trackBackgroundColor as well as some other properties.

Closes #4883

### CODE REVIEW

File: components/Slider/src/MDCSlider.m [hunk 1]
```diff
@@ -209,6 +213,24 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
   return color;
 }
 
+- (void)setThumbRadius:(CGFloat)thumbRadius forState:(UIControlState)state {
+  _thumbRadiusesForState[@(state)] = [NSNumber numberWithFloat:thumbRadius];
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Should we NSAssert if they assign a negative value?

File: components/Slider/src/MDCSlider.m [hunk 2]
```diff
@@ -209,6 +213,24 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
...
+  if (self.state == state) {
+    [self updateThumbForState];
+  }
+}
+
+- (CGFloat)thumbRadiusForState:(UIControlState)state {
+  if ([_thumbRadiusesForState objectForKey:@(state)]) {
+    return [_thumbRadiusesForState[@(state)] floatValue];
+  }
+  CGFloat radius = 0.f;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Can we use the default value instead of 0? I don't think we want clients to be "hiding" the thumb by making it too small to render.

File: components/Slider/tests/unit/SliderTests.m [hunk 3]
```diff
@@ -913,6 +913,42 @@ - (void)testThumbElevationDefault {
   XCTAssertEqual(self.slider.thumbElevation, MDCShadowElevationNone);
 }
 
+- (void)testThumbRadiusForState {
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m:916]: Please test the default/fallback behavior as well.

File: components/Slider/src/MDCSlider.h [hunk 4]
```diff
@@ -161,6 +161,28 @@ IB_DESIGNABLE
...
+
+ In general, if a property is not specified for a state, the default is to use the @c
+ UIControlStateNormal value. If the @c UIControlStateNormal value is not set, then the property
+ defaults to a default value. Therefore, at a minimum, you should set the value for the normal
+ state.
+
+ @param thumbRadius The radius of the thumb.
+ @param state The state of the slider.
+ */
+- (void)setThumbRadius:(CGFloat)thumbRadius forState:(UIControlState)state;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h]: What is happening to `-thumbRadius`?  Does it just get assigned to the `.normal` state?

File: components/Slider/src/MDCSlider.m [hunk 5]
```diff
@@ -229,6 +251,14 @@ - (void)updateColorsForState {
   _thumbTrack.trackOffTickColor = [self backgroundTrackTickColorForState:self.state];
 }
 
+- (void)updateThumbForState {
+  if (!self.isStatefulAPIEnabled) {
+    return;
+  }
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m:256]: I feel as though if the stateful API is disabled, we should restore the ThumbTrack's thumbRadius to the `-thumbRadius` property value.  What do you think?

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m:916]: Please add a test that the legacy behavior is not affected when the API is off.

**Review by romoore — DISMISSED** [review]

File: components/Slider/src/MDCSlider.m (hunk 2 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Done

File: components/Slider/src/MDCSlider.m (hunk 1 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Done

File: components/Slider/src/MDCSlider.h (hunk 4 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.h]: Yes

File: components/Slider/src/MDCSlider.m (hunk 5 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m:256]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m:916]: Done

Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m:916]: I believe this is the same as the comment above.

File: components/Slider/src/MDCSlider.h (hunk 4 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.h]: Please document the behavior in the header. Similar updates to the documentation was added for the other properties affected by the "stateful API" flag.

File: components/Slider/src/MDCSlider.m [hunk 6]
```diff
@@ -229,6 +252,14 @@ - (void)updateColorsForState {
   _thumbTrack.trackOffTickColor = [self backgroundTrackTickColorForState:self.state];
 }
 
+- (void)updateThumbForState {
+  if (!self.isStatefulAPIEnabled) {
+    _thumbTrack.thumbRadius = kSliderDefaultThumbRadius;
+  }
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Won't this be overridden by the stateful API value below?

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (romoore) [reply components/Slider/tests/unit/SliderTests.m:916]: I don't see the test added. Can you please clarify what you mean by "the same comment above"? Do you mean "Done"?

screen shot 2018-08-27 at 3 40 21 pm

**Review by romoore — DISMISSED** [review]

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m:916]: What's the difference between the legacy behavior and fallback behavior?

File: components/Slider/src/MDCSlider.m (hunk 6 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Forgot to add the return statement.

File: components/Slider/src/MDCSlider.h (hunk 4 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.h]: I added documentation in the header for `thumbRadius` property.

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (romoore) [reply components/Slider/tests/unit/SliderTests.m:916]: Sorry, by "legacy behavior" I meant when `.statefulAPIEnabled = NO;`.  Fallback is that each of the UIControl states will use the value for `.normal` if no explicit value is configured for that state.

File: components/Slider/src/MDCSlider.h [hunk 7]
```diff
@@ -172,6 +194,7 @@ IB_DESIGNABLE
  The radius of the cursor (thumb).
 
  Default value is 6 points.
+ @note Has no effect if @c statefulAPIEnabled is @c YES.
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h]: Can we discourage clients from using this? I hope we are planning to deprecate this property in the future. 

Would you be willing to follow the pattern adopted for `trackBackgroundColor`, `color`, and others and document the upcoming deprecation as well as the recommended alternative (the new `thumbRadius:forState:` APIs)?

File: components/Slider/tests/unit/SliderTests.m [hunk 8]
```diff
@@ -913,6 +913,64 @@ - (void)testThumbElevationDefault {
...
+  // Given
+  self.slider.statefulAPIEnabled = YES;
+
+  // When
+  CGFloat defaultThumbRadius = self.slider.thumbRadius;
+  CGFloat thumbRadius = 100.f;
+  self.slider.thumbRadius = thumbRadius;
+
+  XCTAssertEqual(self.slider.thumbRadius, defaultThumbRadius);
+
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m]: Nit: Please remove newline.

File: components/Slider/tests/unit/SliderTests.m [hunk 9]
```diff
@@ -913,6 +913,64 @@ - (void)testThumbElevationDefault {
...
+- (void)testThumbRadiusStatefulAPI {
+  // Given
+  self.slider.statefulAPIEnabled = YES;
+
+  // When
+  CGFloat defaultThumbRadius = self.slider.thumbRadius;
+  CGFloat thumbRadius = 100.f;
+  self.slider.thumbRadius = thumbRadius;
+
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m:960]: Nit: Missing "Then" comment.

File: components/Slider/tests/unit/SliderTests.m [hunk 10]
```diff
@@ -913,6 +913,64 @@ - (void)testThumbElevationDefault {
...
+  self.slider.highlighted = YES;
+  XCTAssertEqual(self.slider.thumbRadius, highlighted);
+
+  self.slider.enabled = YES;
+  self.slider.selected = YES;
+  self.slider.highlighted = NO;
+  XCTAssertEqual(self.slider.thumbRadius, selected);
+}
+
+- (void)testThumbRadiusStatefulAPI {
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m]: Nit/suggestion: `testThumbRadiusHasNoEffectWhenStatefulAPIIsEnabled`

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (romoore) [reply components/Slider/tests/unit/SliderTests.m:916]: I don't see the test for the fallback behavior.  Could you please add a test similar to those in cl/194783040 where each UIControlState is checked?

File: components/Slider/tests/unit/SliderTests.m (hunk 10 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 9 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m:960]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 8 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m]: Done

File: components/Slider/src/MDCSlider.h (hunk 7 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.h]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m:916]: Done

**Comment by jverkoey** [comment]
Is this PR stuck? Either mark as needs changes, label it as blocked, land it, or close it please.

**Comment by codeman7** [comment]
It's just waiting review.

File: components/Slider/src/MDCSlider.h [hunk 11]
```diff
@@ -162,18 +162,33 @@ IB_DESIGNABLE
...
+ In general, if a property is not specified for a state, the default is to use the @c
+ UIControlStateNormal value. If the @c UIControlStateNormal value is not set, then the property
+ defaults to a default value. Therefore, at a minimum, you should set the value for the normal
+ state.
+
+ @param thumbRadius The radius of the thumb.
+ @param state The state of the slider.
  */
-@property(nonatomic, strong, nullable) UIColor *inkColor;
+- (void)setThumbRadius:(CGFloat)thumbRadius forState:(UIControlState)state;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h]: Can we please document what happens with negative or zero values?

File: components/Slider/src/MDCSlider.m [hunk 12]
```diff
@@ -209,6 +213,25 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
   return color;
 }
 
+- (void)setThumbRadius:(CGFloat)thumbRadius forState:(UIControlState)state {
+  NSAssert(thumbRadius >= 0, @"Thumb radius must be a positive value");
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Please add code to handle this case in non-debug code.

File: components/Slider/src/MDCSlider.m [hunk 13]
```diff
@@ -209,6 +213,25 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
...
+  float radius = (float)thumbRadius;
+  _thumbRadiusesForState[@(state)] = [NSNumber numberWithFloat:radius];
+  if (self.state == state) {
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Please update the thumb radius regardless of the current state. Consider what happens if it is currently `.highlighted` and the value for `.normal` was just updated. If we're relying on the fallback behavior for `.highlighted`, then we should see the change immediately.

File: components/Slider/src/MDCSlider.m [hunk 14]
```diff
@@ -209,6 +213,25 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
...
+    [self updateThumbForState];
+  }
+}
+
+- (CGFloat)thumbRadiusForState:(UIControlState)state {
+  if ([_thumbRadiusesForState objectForKey:@(state)]) {
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Please use a local value (`NSNumber`) for the return value so we don't need to perform a dictionary lookup twice.

File: components/Slider/src/MDCSlider.m [hunk 15]
```diff
@@ -209,6 +213,25 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
   return color;
 }
 
+- (void)setThumbRadius:(CGFloat)thumbRadius forState:(UIControlState)state {
+  NSAssert(thumbRadius >= 0, @"Thumb radius must be a positive value");
+  float radius = (float)thumbRadius;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Let's keep this as a CGFloat or change the API to be `float` type. I'd prefer that we retain the full value provided by clients.

File: components/Slider/src/MDCSlider.m [hunk 16]
```diff
@@ -209,6 +213,25 @@ - (UIColor *)backgroundTrackTickColorForState:(UIControlState)state {
...
+  _thumbRadiusesForState[@(state)] = [NSNumber numberWithFloat:radius];
+  if (self.state == state) {
+    [self updateThumbForState];
+  }
+}
+
+- (CGFloat)thumbRadiusForState:(UIControlState)state {
+  if ([_thumbRadiusesForState objectForKey:@(state)]) {
+    return [_thumbRadiusesForState[@(state)] floatValue];
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Please use `doubleValue` and cast to CGFloat.

File: components/Slider/src/MDCSlider.m [hunk 17]
```diff
@@ -229,9 +252,22 @@ - (void)updateColorsForState {
...
+}
+
 #pragma mark - ThumbTrack passthrough methods
 
 - (void)setThumbRadius:(CGFloat)thumbRadius {
+  if (self.statefulAPIEnabled) {
+    return;
+  }
+  float radius = (float)thumbRadius;
+  _thumbRadiusesForState[@(UIControlStateNormal)] = [NSNumber numberWithFloat:radius];
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m]: Let's make this a `numberWithDouble:`

File: components/Slider/tests/unit/SliderTests.m [hunk 18]
```diff
@@ -913,6 +913,80 @@ - (void)testThumbElevationDefault {
   XCTAssertEqual(self.slider.thumbElevation, MDCShadowElevationNone);
 }
 
+- (void)testThumbRadiusForState {
+  // Given
+  self.slider.statefulAPIEnabled = YES;
+  CGFloat normal = 12.f;
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m]: Nit/naming: Let's name these `normalRadius`, `disabledRadius`, *etc.*

File: components/Slider/tests/unit/SliderTests.m [hunk 19]
```diff
@@ -913,6 +913,80 @@ - (void)testThumbElevationDefault {
...
+  self.slider.highlighted = YES;
+  XCTAssertEqual(self.slider.thumbRadius, highlighted);
+
+  self.slider.enabled = YES;
+  self.slider.selected = YES;
+  self.slider.highlighted = NO;
+  XCTAssertEqual(self.slider.thumbRadius, selected);
+}
+
+- (void)testThumbRadiusHasNoEffectWhenStatefulAPIIsEnabled {
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m]: If I turn on stateful API, set the radius, then turn off the stateful API, should it be ignored? For the reverse (stateful off, setRadius, stateful on), updating the radius will affect the stateful API.

File: components/Slider/tests/unit/SliderTests.m [hunk 20]
```diff
@@ -913,6 +913,80 @@ - (void)testThumbElevationDefault {
...
+  XCTAssertEqual(self.slider.thumbRadius, defaultThumbRadius);
+}
+
+- (void)testThumbRadiusDefaultWhenStatefulAPIIsDisabled {
+  // Given
+  self.slider.statefulAPIEnabled = NO;
+
+  // When
+  CGFloat thumbRadius = 40.f;
+  [self.slider setThumbRadius:thumbRadius];
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m]: Please use dot notation for property assignments `self.slider.thumbRadius = thumbRadius;`

**Review by romoore — DISMISSED** [review]
Please add the "fallback" unit tests we talked about for the new stateful API.

File: components/Slider/src/MDCSlider.m (hunk 1 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Have instead decided to just always use the absolute value.

File: components/Slider/src/MDCSlider.h (hunk 11 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.h]: Have documented that if a client uses a negative value then the absolute value will be used.

File: components/Slider/src/MDCSlider.m (hunk 12 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Going to use the absolute value.

File: components/Slider/src/MDCSlider.m (hunk 15 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Done

File: components/Slider/src/MDCSlider.m (hunk 16 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Updated the API to use a `float` instead of a `double` or `CGFloat`.

File: components/Slider/src/MDCSlider.m (hunk 17 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Updated the API to use `float`.

File: components/Slider/tests/unit/SliderTests.m (hunk 18 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 20 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m]: Done

File: components/Slider/src/MDCSlider.m (hunk 13 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Done

File: components/Slider/src/MDCSlider.m (hunk 14 above)
Comment (codeman7) [reply components/Slider/src/MDCSlider.m]: Done

File: components/Slider/tests/unit/SliderTests.m (hunk 19 above)
Comment (codeman7) [reply components/Slider/tests/unit/SliderTests.m]: No, if you set the radius when stateful API is on then that is the thumb radius. If you turn it off later how would we know what it was before setting it? We would have to store a value inside the implementation `nonStatefulThumbRadius` we don't do that for any other properties and that seems *hacky*.

**Comment by codeman7** [comment]
The only test we discussed was added `testThumbRadiusDefaultWhenStatefulAPIIsDisabled`.

File: components/Slider/src/MDCSlider.m [hunk 21]
```diff
@@ -229,9 +250,22 @@ - (void)updateColorsForState {
...
+}
+
 #pragma mark - ThumbTrack passthrough methods
 
 - (void)setThumbRadius:(CGFloat)thumbRadius {
+  if (self.statefulAPIEnabled) {
+    return;
+  }
+  double radius = (double)thumbRadius;
+  _thumbRadiusesForState[@(UIControlStateNormal)] = [NSNumber numberWithDouble:radius];
```
Comment (romoore) [inline components/Slider/src/MDCSlider.m:268]: I don't think that using the old API should affect the new API.  What do you think?  Probably easiest to just assign the `_thumbTrack.thumbRadius`  if the new API is disabled.

File: components/Slider/tests/unit/SliderTests.m [hunk 22]
```diff
@@ -913,6 +913,96 @@ - (void)testThumbElevationDefault {
...
+
+  NSUInteger maxState = UIControlStateNormal | UIControlStateSelected | UIControlStateDisabled |
+  UIControlStateHighlighted;
+
+  for (NSUInteger state = 0; state <= maxState; ++state) {
+    XCTAssertEqual([self.slider thumbRadiusForState:state], thumbRadius);
+  }
+}
+
+- (void)testThumbRadius {
```
Comment (romoore) [inline components/Slider/tests/unit/SliderTests.m:997]: Please add a tests for negative and zero values.

File: components/Slider/tests/unit/SliderTests.m (hunk 3 above)
Comment (romoore) [reply components/Slider/tests/unit/SliderTests.m:916]: Done with `-testThumbRadiusDefaultWhenStatefulAPIEnabled`

File: components/Slider/src/MDCSlider.h [hunk 23]
```diff
@@ -162,18 +162,34 @@ IB_DESIGNABLE
...
+ UIControlStateNormal value. If the @c UIControlStateNormal value is not set, then the property
+ defaults to a default value. Therefore, at a minimum, you should set the value for the normal
+ state.
+
+ @param thumbRadius The radius of the thumb.
+ @param state The state of the slider.
+ @note If a negative value is passed the absolute value will be used.
  */
-@property(nonatomic, strong, nullable) UIColor *inkColor;
+- (void)setThumbRadius:(double)thumbRadius forState:(UIControlState)state;
```
Comment (romoore) [inline components/Slider/src/MDCSlider.h:176]: Any reason why this cannot be `CGFloat` at the API? It would match the legacy API's type and using explicit `double` types in UI APIs is pretty unusual on iOS.

File: components/Slider/src/MDCSlider.m (hunk 5 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.m:256]: Given what you needed to do (affect the new API from the old API), I think this might have been a mistake. I was thinking that an iVar could be used, but it may just be better to go back to the original version.  When we spoke offline about why a client would be turning the API on and off, I think you made a good point.  We can probably expect clients to only go from the legacy to the new API and not back again.

**Review by romoore — CHANGES_REQUESTED** [review]
